components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt

```diff
@@ -173,45 +173,85 @@ class Nimbus(
     override fun updateExperiments() {
         scope.launch {
             try {
-                nimbus.updateExperiments()
-
-                // Get the experiments to record in telemetry
-                nimbus.getActiveExperiments().let {
-                    if (it.any()) {
-                        recordExperimentTelemetry(it)
-                        // The current plan is to have the nimbus-sdk updateExperiments() function
-                        // return a diff of the experiments that have been received, at which point we
-                        // can emit the appropriate telemetry events and notify observers of just the
-                        // diff
-                        notifyObservers { onUpdatesApplied(it) }
-                    }
+                nimbus.updateExperiments().also { enrollmentChangeEvents ->
+                    // Record any events in telemetry
+                    recordExperimentTelemetryEvents(enrollmentChangeEvents)
+                }
+                // Update the active experiment annotations in Glean
+                nimbus.getActiveExperiments().also { experiments ->
+                    recordExperimentTelemetry(experiments)
+                    notifyObservers { onUpdatesApplied(experiments) }
                 }
             } catch (e: ErrorException.RequestError) {
                 logger.info("Error fetching experiments from endpoint: $e")
             } catch (e: ErrorException.InvalidExperimentResponse) {
                 logger.info("Invalid experiment response: $e")
+            } catch (e: ErrorException) {
+                logger.info("Nimbus Error: $e")
             }
         }
     }
 
     override fun optOut(experimentId: String) {
-        nimbus.optOut(experimentId)
+        nimbus.optOut(experimentId).also { enrollmentChangeEvents ->
+            recordExperimentTelemetryEvents(enrollmentChangeEvents)
+        }
+
+        // Set the experiment inactive in the telemetry annotations
+        Glean.setExperimentInactive(experimentId)
     }
 
     // This function shouldn't be exposed to the public API, but is meant for testing purposes to
     // force an experiment/branch enrollment.
     @VisibleForTesting(otherwise = VisibleForTesting.NONE)
     internal fun optInWithBranch(experiment: String, branch: String) {
-        nimbus.optInWithBranch(experiment, branch)
+        nimbus.optInWithBranch(experiment, branch).also { enrollmentChangeEvents ->
+            recordExperimentTelemetryEvents(enrollmentChangeEvents)
+        }
+
+        // If we were successfully enrolled in the experiment (it existed), then we can update
+        // telemetry annotation with the experiment and branch values.
+        nimbus.getExperimentBranch(experiment)?.also { branchName ->
+            Glean.setExperimentActive(experiment, branchName)
+        }
```

Comment:
I don't think we can get this information from the `enrollmentChangeEvents` because it only shows the 'enrolled' experiment once so as to only trigger the enrollment event once per client per experiment.  The events in the list are only there to indicate a change of state in the experiment and don't include active experiments whose enrollment state has not changed.

Reply:
As for recording telemetry?  If we opt-in for testing, they may be using the Glean debug tools to verify the enrollment or events, so I think it's still a good idea to update the annotations.  Glean shouldn't know or care if we are testing, we could further annotate the ping to indicate that the user "opted-in" here also.

Reply:
On the other hand, it may be useful to do this to test that the expected telemetry is being emitted. We'll also end up calling `Glean.setExperimentActive` on the opted-in experiment on next startup anyway, so I'm not sure it would have much effect if we avoided calling it here.